crates/services/block_aggregator_api/src/api/protobuf_adapter.rs

```diff
@@ -152,41 +188,97 @@ impl BlockAggregator for Server {
-        let addr = url.parse().unwrap();
-        let (shutdown_sender, shutdown_receiver) = tokio::sync::oneshot::channel::<()>();
-        let _server_task_handle = tokio::spawn(async move {
-            let service = tonic::transport::Server::builder()
-                .add_service(ProtoBlockAggregatorServer::new(server));
-            tokio::select! {
-                res = service.serve(addr) => {
+pub struct ServerTask {
+    addr: std::net::SocketAddr,
+    query_sender:
+        tokio::sync::mpsc::Sender<BlockAggregatorQuery<BlockRangeResponse, ProtoBlock>>,
+    router: Option<Router>,
+}
+#[async_trait::async_trait]
+impl RunnableService for ServerTask {
+    const NAME: &'static str = "ProtobufServerTask";
+    type SharedData = ();
+    type Task = Self;
+    type TaskParams = ();
+
+    fn shared_data(&self) -> Self::SharedData {}
+
+    async fn into_task(
+        mut self,
+        _state_watcher: &StateWatcher,
+        _params: Self::TaskParams,
+    ) -> anyhow::Result<Self::Task> {
+        self.start_router()?;
+        Ok(self)
+    }
+}
+
+impl ServerTask {
+    fn start_router(&mut self) -> anyhow::Result<()> {
+        let server = Server::new(self.query_sender.clone());
+        let router = tonic::transport::Server::builder()
+            .add_service(ProtoBlockAggregatorServer::new(server));
+        self.router = Some(router);
+        Ok(())
+    }
+
+    fn get_router(&mut self) -> anyhow::Result<Router> {
+        self.router
+            .take()
+            .ok_or_else(|| anyhow!("Router has not been initialized yet"))
+    }
+}
+
+impl RunnableTask for ServerTask {
+    async fn run(&mut self, watcher: &mut StateWatcher) -> TaskNextAction {
+        let router_res = self.get_router();
+        let router = try_or_stop!(router_res, |e| tracing::error!(
+            "Failed to get router, has not been started: {:?}",
+            e
+        ));
+        tokio::select! {
+                res = router.serve(self.addr) => {
                     if let Err(e) = res {
                         tracing::error!("BlockAggregator tonic server error: {}", e);
+                        TaskNextAction::ErrorContinue(anyhow!(e))
                     } else {
                         tracing::info!("BlockAggregator tonic server stopped");
+                        TaskNextAction::Stop
                     }
                 },
-                _ = shutdown_receiver => {
-                    tracing::info!("Shutting down BlockAggregator tonic server");
-                },
+            _ = watcher.while_started() => {
+                TaskNextAction::Stop
             }
+        }
+    }
+
+    async fn shutdown(self) -> anyhow::Result<()> {
+        Ok(())
+    }
+}
+
+impl ProtobufAPI {
+    pub fn new(url: String) -> Result<Self> {
+        let (query_sender, query_receiver) = tokio::sync::mpsc::channel::<
+            BlockAggregatorQuery<BlockRangeResponse, ProtoBlock>,
+        >(100);
```

Comment:
Looks like `100` should be customizable via CLI, becuase it defiend how many concurrent user we can support via RPC